File: Rust/2020/04.rs
#![feature(test)]

use std::collections::HashMap;

use regex::Regex;

type Input = Vec<HashMap<String, String>>;

fn setup(input: &str) -> Input {
    let mut out = Vec::new();
    for block in input.split("\n\n") {
        let mut cur = HashMap::new();
        for line in block.trim().lines() {
            for (key, value) in line.split(' ').map(|a| {
                let mut x = a.trim().split(':');
                (x.next().unwrap(), x.next().unwrap())
            }) {
                cur.insert(key.to_string(), value.to_string());
            }
        }
        out.push(cur);
    }
    out
}

fn part1(input: &Input) -> String {
    let mut out = 0;
    for passport in input {
        if ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]
            .iter()
            .all(|x| passport.contains_key(&x.to_string()))
        {
            out += 1;
        }
    }
    out.to_string()
}

type Rule = (&'static str, fn(&String) -> bool);

const RULES: [Rule; 7] = [
    ("byr", |x: &String| {
        let num: i32 = x.parse().unwrap();
        (1920..=2002).contains(&num)
    }),
    ("iyr", |x: &String| {
        let num: i32 = x.parse().unwrap();
        (2010..=2020).contains(&num)
    }),
    ("eyr", |x: &String| {
        let num: i32 = x.parse().unwrap();
        (2020..=2030).contains(&num)
    }),
    ("hgt", |x: &String| {
        let regex = Regex::new(r"^(\d+)(cm|in)$").unwrap();
        match regex.captures(x) {
            None => false,
            Some(result) => {
                let num: i32 = result[1].parse().unwrap();
                match &result[2] {
                    "cm" => (150..=193).contains(&num),
                    "in" => (59..=76).contains(&num),
                    _ => false,
                }
            }
        }
    }),
    ("hcl", |x: &String| {
        let regex = Regex::new(r"^#[0-9a-f]{6}$").unwrap();
        regex.captures(x).is_some()
    }),
    ("ecl", |x: &String| {
        ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"].contains(&x.as_str())
    }),
    ("pid", |x: &String| {
        let regex = Regex::new(r"^\d{9}$").unwrap();
        regex.captures(x).is_some()
    }),
];

fn part2(input: &Input) -> String {
    let mut out = 0;
    for passport in input {
        if RULES
            .iter()
            .all(|(key, validator)| match passport.get(&key.to_string()) {
                None => false,
                Some(value) => validator(value),
            })
        {
            out += 1;
        }
    }
    out.to_string()
}

aoc::main!(2020, 4, ex: 1[a]);

File: Python/2020/04.py
from lib import *

input = read_input(2020, 4)


out = 0
for kv in map(str.split, input.split("\n\n")):
    x = {x.split(":")[0] for x in kv}
    out += all(e in x for e in ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"])

print(out)


def valid(x):
    if not re.match(r"^\d{4}$", x.get("byr", "")) or not 1920 <= int(x["byr"]) <= 2002:
        return False

    if not re.match(r"^\d{4}$", x.get("iyr", "")) or not 2010 <= int(x["iyr"]) <= 2020:
        return False

    if not re.match(r"^\d{4}$", x.get("eyr", "")) or not 2020 <= int(x["eyr"]) <= 2030:
        return False

    if m := re.match(r"^(\d+)(in|cm)$", x.get("hgt", "")):
        a, b = m.groups()

        if b == "cm" and not 150 <= int(a) <= 193:
            return False

        elif b == "in" and not 59 <= int(a) <= 76:
            return False

    else:
        return False

    if not re.match(r"^#[\da-f]{6}$", x.get("hcl", "")):
        return False

    if x.get("ecl", "") not in "amb blu brn gry grn hzl oth".split():
        return False

    return bool(re.match(r"^\d{9}$", x.get("pid", "")))


out = 0
for kv in map(str.split, input.split("\n\n")):
    out += valid(dict(x.split(":") for x in kv))

print(out)
